File: proiectarea algoritmilor/Balan Bianca-Ioana/flipping sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void flip(int arr[], int k) {
    for (int i = 0; i < k / 2; i++)
        swap(arr[i], arr[k - i - 1]);
}
int findMaxIndex(int arr[], int n) {
    int maxIdx = 0;
    for (int i = 1; i < n; i++)
        if (arr[i] > arr[maxIdx])
            maxIdx = i;
    return maxIdx;
}
void pancakeSort(int arr[], int n) {
    for (int currSize = n; currSize > 1; currSize--) {
        int maxIdx = findMaxIndex(arr, currSize);
        if (maxIdx != currSize - 1) {
            if (maxIdx > 0) {
                flip(arr, maxIdx + 1);
                cout << "Flip pana la index " << maxIdx << ": ";
                for (int i = 0; i < n; i++) cout << arr[i] << " ";
                cout << endl;
            }
            flip(arr, currSize);
            cout << "Flip pana la index " << currSize - 1 << ": ";
            for (int i = 0; i < n; i++) cout << arr[i] << " ";
            cout << endl;
        }
    }
}
int main() {
    int arr[] = {3, 6, 1, 10, 7, 5};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Vectorul initial: ";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;

    pancakeSort(arr, n);

    cout << "Vectorul sortat: ";
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;

    return 0;
}

File: Miclaus Catalin/flipping.sort.py
    """inverseaza ordindef flip(arr, k):
ea"""
    arr[:k] = arr[:k][::-1]

def find_max_index(arr, n):
    """Finds the index of the maximum element in the first n elements."""
    max_idx = 0
    for i in range(1, n):
        if arr[i] > arr[max_idx]:
            max_idx = i
    return max_idx

def pancake_sort(arr):
    """Sorts the array using the pancake flipping sort algorithm."""
    n = len(arr)
    for curr_size in range(n, 1, -1):
        # Find the maximum element in the current unsorted part of the array
        max_idx = find_max_index(arr, curr_size)
        
        # If the maximum element is not at its correct position
        if max_idx != curr_size - 1:
            # Bring the maximum element to the beginning if necessary
            if max_idx != 0:
                flip(arr, max_idx + 1)
            
            # Bring the maximum element to its correct position
            flip(arr, curr_size)
    return arr

# Example usage
arr = [3, 6, 2, 4, 8, 1, 5, 7]
print("Original array:", arr)
sorted_arr = pancake_sort(arr)
print("Sorted array:", sorted_arr)
